### ISSUE TITLE
Feature request: Add support for esm.

### ISSUE THREAD

**Issue by coreyfarrell** [gulpjs/interpret#52]
I'd like to request that esm be added if possible.  Bad news I just tried adding it and couldn't get it to work.  I added `'.esm.mjs': 'esm',` to extensions and `'.esm.mjs'` to jsVariantExtensions.  This caused the following from gulp:
```
$ gulp
[21:05:15] Requiring external module esm
/home/coreyfarrell/test/gulpfile.esm.mjs:1
(function (exports, require, module, __filename, __dirname) { import log from 'gulplog';
                                                                     ^^^

SyntaxError: Unexpected identifier
    at new Script (vm.js:74:7)
    at createScript (vm.js:246:10)
    at Object.runInThisContext (vm.js:298:10)
    at Module._compile (internal/modules/cjs/loader.js:670:28)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:713:10)
    at Module.load (internal/modules/cjs/loader.js:612:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:551:12)
    at Function.Module._load (internal/modules/cjs/loader.js:543:3)
    at Module.require (internal/modules/cjs/loader.js:650:17)
    at require (internal/modules/cjs/helpers.js:20:18)
```
The result was the same for `gulp` and `npx gulp` (global gulp-cli vs local gulp-cli).

**Comment by phated** [comment]
Thanks for opening the issue! I think the implementation will be closer to Babel (with the custom function wrapper) but haven't looked into it yet.

**Comment by phated** [comment]
@coreyfarrell I'm digging into this more but it seems like it'd require changes on the `esm` side because they expect the consumer to use a returned `require`-style function but we expect things to be registered onto a extension.

**Comment by phated** [comment]
Actually, I think I found the answer at https://github.com/standard-things/esm/issues/433

**Comment by phated** [comment]
Need to do some testing on the changes using `rechoir` before I publish.

**Comment by coreyfarrell** [comment]
I just did some local testing against gulp, one issue I ran into it doesn't seem to respect any of the ways to set [ESM options](https://github.com/standard-things/esm/#options).  To enable debug I had to edit interpret to use `var esmLoader = hook(module, {debug: true});`.

Note I'm looking into debug to find out why `import {rollup} from 'rollup';` doesn't work.  I think it's an esm bug, I'm able to reproduce this issue with a regular `node -r esm script.mjs`.

**Comment by phated** [comment]
@coreyfarrell I'm guessing they don't look up options when it's used programmatically - might want to check that too.

**Comment by coreyfarrell** [comment]
I got an [explanation](https://github.com/standard-things/esm/issues/485#event-1709556810) of the issue I was having with rollup.  I had to drop my plan to use a dual mode `main` file in my own package, instead using `"main": "index.js", "module": "index.mjs"`.

At this point I'm not sure how important it is to be able to set the ESM options, I'm able to use `gulpfile.esm.js` with the default options.

**Comment by coreyfarrell** [comment]
@phated Would you be willing to publish an updated version that includes `.esm.js`?  I've been using this for a while and once I understood the limitations of the `esm` module it works great for me.  Even if I'm pointing to a `@next` version I'd prefer to use an npm based dependency over `gulpjs/interpret#master`.

**Comment by phated** [comment]
Yeah, I can probably get something published this week.

**Comment by phated** [comment]
@coreyfarrell potentially great news, I'm just wrapping up a complete test suite for this project and should have the ESM update published as soon as my Window's CI succeeds.

**Comment by phated** [comment]
@coreyfarrell just published as 1.2.0!!
